handshake/crypto_setup_client.go

```diff
@@ -354,9 +377,13 @@ func (h *cryptoSetupClient) DiversificationNonce() []byte {
 }
 
 func (h *cryptoSetupClient) SetDiversificationNonce(data []byte) error {
+	h.mutex.Lock()
+	defer h.mutex.Unlock()
+
 	if len(h.diversificationNonce) == 0 {
 		h.diversificationNonce = data
-		return h.maybeUpgradeCrypto()
+		h.divNonceChan <- struct{}{}
```

Comment:
This can deadlock if the chan is not being read (e.g. if we received an invalid handshake message and the loop terminated).